### ISSUE TITLE
stereo RGB camera 3D reconstruction

### ISSUE THREAD

**Issue by max-zilla** [terraref/computing-pipeline#467]
We need to figure out status of this effort. 

- Zongyang has looked at this in Season 2 with these chessboard images (left/right)
![image](https://user-images.githubusercontent.com/14166971/41249801-b8b5f60a-6d7a-11e8-8597-54bd30ff5eb4.png)
![image 1](https://user-images.githubusercontent.com/14166971/41249807-bdd8e2d2-6d7a-11e8-9783-98fb7129ebdf.png)

- After Season 2, something changed and the previous calibration data cannot be used
- @ZongyangLi do we know what changed? @NewcombMaria said the cameras were not moved for stereoRGB.

- The later seasons don't work - maybe aperture changed? Should not affect

- We need to get new calibration parameters/chessboard images for current season and determine status of this work.

**Comment by max-zilla** [comment]
This seems to be a follow-on from this long-stale issue:
https://github.com/terraref/computing-pipeline/issues/355

**Comment by max-zilla** [comment]
@ZongyangLi can you explain a bit of what you would need to resume work on this for season 6?
- new checkerboard images?
- metadata (e.g. lens settings)
- something else?

**Comment by max-zilla** [comment]
Chessboard images are needed for both sides of camera.

**Comment by JeffWhiteAZ** [comment]
Do we have anything approaching a standard operating procedure (SOP) for the stereo cameras?
If I understand correctly, each image should be taken with the board in a different position. Should the angle from the vertical be varied or kept fairly constant?

**Comment by ZongyangLi** [comment]
@JeffWhiteAZ I think you are correct, each image should be taken with the board in a different position, but the angle from the vertical should keep constant.

**Comment by NewcombMaria** [comment]
@ZongyangLi @max-zilla we collected calibration images of the black/white checkerboard this morning.  Jeffrey D. wrote a script to take 25 images and we moved the checkerboard between images.  There's also a meter stick in each image.  The name of the scan is 'S6_Sorghum_2018_Calibration_Checkerboard_RGB_IR.cs', start time 08:46.  Examples (png conversions) copied below.

Let us know if these images work for the stereo calibration.  We can collect more or different images as needed. 
![image](https://user-images.githubusercontent.com/22008637/41992591-c2458156-79fd-11e8-8629-80de955e9a9a.png)
![image](https://user-images.githubusercontent.com/22008637/41992611-d324306c-79fd-11e8-8ede-24751d231314.png)

**Comment by ZongyangLi** [comment]
@NewcombMaria Thank you very much for the fast respond of calibration images.

I have some little concerns about the images:
1. The chessboard seems not in a flat plane, in the second example, the left top corner is a little bit low.
2. Some corner is covered by leaves or some other boundaries, this might affect the corner extraction position.

**Comment by NewcombMaria** [comment]
@ZongyangLi we are limited in options for placing targets now because the plants are so tall.  We only have one target table that is about 40cm wide - a ladder is needed now for me to reach the top of the table.  When the checkerboard is rotated the corners fall off the edge of the table.  

If needed I can cut leaves to get them out of the way, and if needed I can try to mount it on a board.  There are 25 images total, many with all corners on the table.   We thought you might have enough that are useful.  Let me know after you review the 25 images if you need me to make adjustments and take more.

**Comment by ZongyangLi** [comment]
@NewcombMaria Sure! I will take a look at all 25 images and see if it works.

**Comment by NewcombMaria** [comment]
Thanks @ZongyangLi.  We can mount the checkerboard on solid board so the full surface is rigid.  Any other guidance is welcome before we collect more images.

**Comment by NewcombMaria** [comment]
@ZongyangLi I collected a better set of checkerboard images on 2018-06-29 between 08:51 and 09:01.  The checkerboard is mounted on a black platform.  There's also a set from earlier in the morning (08:04 to 08:13) but the checkerboard is in partial sun from the east side so the images may not be useful.  Here are a couple examples from the second set (~25 images per set)
![image](https://user-images.githubusercontent.com/22008637/42135113-88d4b382-7cfb-11e8-828d-acfb1d4cd46e.png)

![image](https://user-images.githubusercontent.com/22008637/42135121-95cc23d6-7cfb-11e8-92f8-d3b4df144e77.png)

**Comment by ZongyangLi** [comment]
@NewcombMaria Thank you, I think it's good for the camera calibration now.

**Comment by ZongyangLi** [comment]
@NewcombMaria I am sorry I have to pin you here again, do we know how big the squares on the chessboard are?

**Comment by NewcombMaria** [comment]
@ZongyangLi, no problem at all.  Square measurements:  5.525 cm x 5.525 cm.  Width of 8 squares across = 44.2 cm.

**Comment by ZongyangLi** [comment]
Thank you!

**Comment by dlebauer** [comment]
Please document this!

**Comment by NewcombMaria** [comment]
@ZongyangLi and/or @max-zilla, in our local group meeting today there was interest in comparing data from stereo RGB camera 3D reconstruction to data derived from other sensors for height estimates.  Is there an update on the status of the stereo RGB reconstruction?  Was the calibration update successful?  Let us know if you need any additional tests with the stereo cameras.  Thanks.

**Comment by ZongyangLi** [comment]
@NewcombMaria The calibration data has the problem that it was all taken at the same distance from the camera, so the calibration process is unstable. Since the cameras are the same I guess (like, the lenses, and the zoom), we can still calibrate the relative position of the cameras, but using the intrinsic parameters from before.

By using the intrinsic parameters from before, we generated new calibration parameters for current stereo image pairs, the error between left/right image seems good to me, but when I use it to compute disparity image, it's not reasonable to me for now. I think there is still something wrong here may cause by my code or data organization or calibration data, I am still on my way of debugging.

**Comment by jdemieville-ua** [comment]
@ZongyangLi : For the varying distances between camera and target, would you also need the full checkerboard at various angles? I'd ran a test on 2018-08-29 taking RGB images at various heights (1.8-2.75 m, in 5 cm increments) to get an idea of the effect on focus, but the checkerboard has many of the same problems as the first attempt in this Github issue.

A compiled ZIP file for that test with original and processed images is already available on Google Drive (~2.21 GB, [https://drive.google.com/open?id=1gtmjZm8EuDm36emwjCn4a25D32Isr94b](url)). 

If this is insufficient, it wouldn't be too hard at all to add in the repeated images on a timer. I'd just need to know what distances (between camera window and target) you'd like to see and how many repetitions you'd need.

**Comment by ZongyangLi** [comment]
After a frame by frame check, the calibration image pairs from 2018-06-29 now gives me reasonable calibration parameters.

We first find all corners both in left/right images:
![2_left](https://user-images.githubusercontent.com/20230686/45780279-d13ac600-bc22-11e8-8e23-99a04493e45c.jpg)
![2_right](https://user-images.githubusercontent.com/20230686/45780288-d435b680-bc22-11e8-8eb5-d5c190ddb4f0.jpg)
Then compute calibration parameters from 24 pairs of given image pairs(some pairs will ruin the calibration, that's why we can not generate reasonable disparity image before).

By new parameters, we will be able to generate rectify images:
![rectify0](https://user-images.githubusercontent.com/20230686/45780550-6a69dc80-bc23-11e8-8ac0-12d6902794f3.png)
![rectify1](https://user-images.githubusercontent.com/20230686/45780557-6f2e9080-bc23-11e8-9933-009f5564ca91.png)
then use rectify image to compute disparity image:
![disp](https://user-images.githubusercontent.com/20230686/45780591-81103380-bc23-11e8-9942-e7615b0d91fb.png)

By using chessboard images from 2018-06-29, I am able to generate good disparity image for season 6, at least from 2018-05-20 to 2018-07-26.
disparity images from 2018-05-28 and 2018-07-26, brighter pixel means closer to the camera:
![disp](https://user-images.githubusercontent.com/20230686/45780930-58d50480-bc24-11e8-8cdf-5bb42b5130bc.png)
![disp](https://user-images.githubusercontent.com/20230686/45780937-6094a900-bc24-11e8-8cb8-48c1e4da5770.png)


Then I check other seasons. Here are some disparity image from 2017-07-27, 2018-03-20, and 2018-09-18:
![disp](https://user-images.githubusercontent.com/20230686/45780886-39d67280-bc24-11e8-9c0c-35ee7ce365c4.png)
![disp](https://user-images.githubusercontent.com/20230686/45780976-7609d300-bc24-11e8-9b5a-896168911474.png)

Those disparity images from dates other than season 6 is not good for height estimate. I guess the stereo camera was moved by accident during different seasons. A good news is that we may capture new calibration image pairs for new season, since images from 2018-08-29 may not good for a calibration.

**Comment by JeffWhiteAZ** [comment]
Do we have an assessment for each season, including wheat crops, as to whether we can obtain useful height information? Presumably, we have access to and can apply the original calibrations and then determine when they were no longer correct/usable.
In Maricopa, we have been telling people that one of the strengths of the field scanner is that we can compare height data from two instruments-- that being part of the concept of this being a "reference" system. we have also planned manuscript on the assumption that we could compare 3-D reconstructions from the laser scanner and the stereo-RGB.

**Comment by JeffWhiteAZ** [comment]
Jeffrey Demieville asked for clarification on the best way to provide calibration data. We do not have an SOP, and there seems to be confusion over whether the checkerboard pattern needs to be rotated or actually set on an incline. The target we have is printed on plasticized cloth (or something like that). To mount it on an incline, we would need a rigid support. 
The SOP might consider:
- How often the calibration should be conducted (start and end of each season?)
- How many images are needed?
- Whether illumination is critical (preference for sunlit vs shaded)?
- Best type of target and whether it needs to be inclined or rotated

**Comment by ZongyangLi** [comment]
How to get those calibration data:

1. Before starting, make sure we have a large checkerboard (7x7) with known dimensions and known squares size.NOTE: Checkerboard size refers to the number of internal corner
2. Moving the checkerboard, in order to get a good calibration you will need to move the checkerboard around in the camera frame such that:
   a) the checkerboard is detected at the left and right edges of the field of view (X calibration)
   b) the checkerboard is detected at the top and bottom edges of the field of view (Y calibration)
   c) the checkerboard is detected at various angles to the camera ("Skew")
   d) the checkerboard fills the entire field of view (Size calibration)
   e) checkerboard tilted to the left, right, top and bottom (X,Y, and Size calibration)

**Comment by JeffWhiteAZ** [comment]
If the set of images are of good quality in terms of your specifications above, how often would calibrations need to be checked? Based on our past experience, "whenever the camera is moved" isn't safe.

**Comment by ZongyangLi** [comment]
Other SOP Points:

1. How often the calibration should be conducted (start and end of each season?)
    I think start and end of each season should be fine, since no one is moving those cameras.
2. How many images are needed?
    A minimum coverage of above moving points will be 10 image pairs.
3. Whether illumination is critical (preference for sunlit vs shaded)?
   The only critical is all the corners can be detected.
4. Best type of target and whether it needs to be inclined or rotated
   Hard target, points is not moving during calibration.Yes it needs to be inclined or rotated.

**Comment by JeffWhiteAZ** [comment]
Thanks! Very helpful feedback. I'll discuss with Jeffrey D. about creating an SOP, getting or making a better target, etc.

**Comment by jdemieville-ua** [comment]
@ZongyangLi : Does your method require us to use only one checkerboard? I'm curious if we can make a target with multiple checkerboards (labeled for differentiation) that is big enough to fill the frame instead of moving on X-Y. 
A table mounted to a camera tripod or a ball mount would allow us to change tilt, skew, and pitch. 

For point 2: is that 10 image pairs per position (e.g. 10 pairs with various X skew, 10 pairs with various Y pitch, 10 pairs at various tilt positions, etc.) or 10 image pairs total at varying positions?

**Comment by ZongyangLi** [comment]
@jdemieville-ua Yes please use only one checkerboard. The square size in the checkerboard affect the calibration.

At least 10 image pairs at varying positions. To prevent any uncomfortable, I will make 20 pairs in case that some images may not in both scope.

**Comment by jdemieville-ua** [comment]
@ZongyangLi : @NewcombMaria and I collected another set of 25 image pairs for you. We used the same board as before. It was placed both flat on our target table as well as at varying skew, pitch, and tilt. The data is available on Globus at /ua-mac/raw_data/stereoTop/2018-10-25/ with timestamps 15:09-15:19. 

![f761b6dc-483c-49c3-8b96-830636a56ba3_total_resized](https://user-images.githubusercontent.com/40777832/47580314-7a9c6800-d903-11e8-979f-3e30c2ac6aac.png)

**Comment by ZongyangLi** [comment]
@jdemieville-ua @NewcombMaria 

I had created the new calibration parameters for 2018-10-25, it works pretty well. And here are some output from this extractor:
![disp](https://user-images.githubusercontent.com/20230686/47583574-81d06f80-d91d-11e8-926c-63e51c4b0d75.png)
![left](https://user-images.githubusercontent.com/20230686/47583577-84cb6000-d91d-11e8-9bff-130e23a2224b.jpg)

So now the only remaining thing for this extractor is to deploy it on NCSA.

**Comment by max-zilla** [comment]
Excellent. Created followup issue for deployment and closing this one:
https://github.com/terraref/computing-pipeline/issues/529

**Comment by jdemieville-ua** [comment]
@ZongyangLi : @NewcombMaria and I collected an additional set of checkerboard images yesterday with a newer, silicone board (2.25" squares) adhered flat to a piece of plywood. They can be found at /ua-mac/raw_data/stereoTop/2019-02-13/ from 13:31-13:42.
